torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,215 @@
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/freeze_module.h>
+
+#include <torch/csrc/jit/graph_executor_impl.h>
+#include <torch/csrc/jit/passes/inliner.h>
+
+#include <stack>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+class AttributePropagator {
+ private:
+  // Contains attributes that can't be folded or user directs to keep them.
+  std::map<script::Module*, std::set<std::string>> preservedAttrs_;
+
+  void insertModifiedAttr(script::Module& module, std::string name) {
+    auto it = preservedAttrs_.find(&module);
+    if (it != preservedAttrs_.end()) {
+      it->second.insert(name);
+    } else {
+      std::set<std::string> names;
+      names.insert(name);
+      preservedAttrs_[&module] = names;
+    }
+  }
+
+  // We can use a more efficient algorithm to hash each constant GetAttr to its
+  // conrresponding Value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
+      Node* input,
+      std::string m_name,
+      script::Module& attr_module) {
+    std::stack<std::string> names;
+
+    while (!attr_module.hasattr(m_name)) {
+      if (input->kind() == prim::GetAttr) {
+        m_name = input->s(attr::name);
+        names.push(m_name);
+        input = input->inputs()[0]->node();
+      } else {
+        // TODO: handle TupleUnpack.
+        return false;
+      }
+    }
+
+    while (!names.empty()) {
+      auto m_name = names.top();
+      names.pop();
+      attr_module = attr_module.attr(m_name).toModule();
+      auto it = preservedAttrs_.find(&attr_module);
+      if (it != preservedAttrs_.end()) {
+        if (it->second.count(m_name)) {
+          // Bail because this attribute is mutable!
+          return false;
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(&attr_module);
+    return it == preservedAttrs_.end() || !it->second.count(m_name);
+  }
+
+  void recordMutableAttrs(
+      std::shared_ptr<Graph>& graph,
+      const script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr) {
+          auto name = n->s(attr::name);
+          auto attr_module = module_clone;
+          auto input = n->inputs()[0]->node();
+          if (!findConstantAttr(input, name, attr_module)) {
+            continue;
+          }
+          insertModifiedAttr(attr_module, name);
+        }
+      }
+    }
+  }
+
+  std::set<std::string> getReferencedAttrs(
+      std::shared_ptr<Graph>& graph,
+      script::Module& module_clone) {
```

Comment:
Most of the methods in this class have these two parameters. Would it make sense to add them as a class fields and initialize once in the beginning (I'm pretty sure the `module_clone` never changes across these calls)?

Reply:
I made the changes to embed it in class.